File: clients/java/src/main/java/de/neoskop/aviation/AviationClientBuilder.java
package de.neoskop.aviation;

import javax.servlet.http.HttpServletRequest;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Supplier;

/**
 * @author Arne Diekmann
 * @since 31.05.17
 */
public class AviationClientBuilder {
	private String endpointUrl;
	private String token;
	private Map<String, Object> meta = new HashMap<>();
	private Supplier<HttpServletRequest> requestSupplier;
	private boolean realtime = true;

	public static AviationClientBuilder aviation() {
		return new AviationClientBuilder();
	}

	public AviationClientBuilder endpoint(String endpointUrl) {
		this.endpointUrl = endpointUrl;
		return this;
	}

	public AviationClientBuilder token(String token) {
		this.token = token;
		return this;
	}

	public AviationClientBuilder meta(String key, Object value) {
		meta.put(key, value);
		return this;
	}

	/**
	 * Determines whether to update features in real time (defaults to true)
	 * @return itself
	 */
	public AviationClientBuilder realtime(boolean realtime) {
		this.realtime = realtime;
		return this;
	}

	public AviationClientBuilder request(Supplier<HttpServletRequest> requestSupplier) {
		this.requestSupplier = requestSupplier;
		return this;
	}

	public AviationClient mix() {
		return new AviationClient(endpointUrl, token, meta, requestSupplier, realtime);
	}
}

File: clients/typescript/src/aviation-client-builder.ts
import {AviationClient} from "./aviation-client";

export function aviation() {
    return new AviationClientBuilder();
}

export class AviationClientBuilder {
    private _endpointUrl: string;
    private _token: string;
    private _meta: any = {};
    private _cacheTime: number = 3600;
    private _webSocketEnabled: boolean = true;

    public endpoint(endpointUrl: string): AviationClientBuilder  {
        this._endpointUrl = endpointUrl;
        return this;
    }

    public token(token: string): AviationClientBuilder {
        this._token = token;
        return this;
    }

    public cache(cacheTime: number): AviationClientBuilder {
        this._cacheTime = cacheTime;
        return this;
    }

    public meta(meta: any): AviationClientBuilder {
        this._meta = meta;
        return this;
    }

    public websocket(webSocketEnabled: boolean): AviationClientBuilder {
        this._webSocketEnabled = webSocketEnabled;
        return this;
    }

    public mix(): AviationClient {
        return new AviationClient(this._endpointUrl, this._token, this._cacheTime, this._meta, this._webSocketEnabled);
    }
}